File: mobile/android/app/src/main/java/com/quickin/app/AITravelChatService.kt
package com.quickin.app

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

/** One turn in the travel-concierge conversation (`role` is "user" or "assistant"). */
data class AiMessage(val role: String, val content: String)

/**
 * Streaming client for the public AI **travel concierge** endpoint. Mirrors
 * [NotificationService] (HttpURLConnection + org.json on Dispatchers.IO, no
 * third-party HTTP/JSON libs) but reads a **Server-Sent Events** body instead of
 * a single JSON document, emitting each token as it arrives.
 *
 *   POST {base}/api/local/ai/chat   { messages: [{role, content}, …] }   (no auth)
 *     -> text/event-stream:
 *          data: {"delta":"…"}\n\n   (repeated; append delta to the reply)
 *          data: [DONE]\n\n          (once, at the end)
 *          data: {"error":"…"}\n\n   (on a mid-stream failure, then close)
 *     A non-200 may instead return JSON {"error":"…"} (e.g. 503 when the AI key
 *     isn't configured) — surfaced as an [HttpError] the UI shows as a friendly note.
 */
object AITravelChatService {

    /** Thrown on a non-2xx response (or a mid-stream `{"error":…}`) so the UI can show it inline. */
    class HttpError(val code: Int, message: String) : RuntimeException(message)

    /**
     * Opens the SSE stream for [messages] and invokes [onDelta] once per token
     * (already hopped back to the caller's context is the caller's job — this runs
     * the network read on Dispatchers.IO and calls [onDelta] from there). Returns
     * normally on `[DONE]`; throws [HttpError] on a non-2xx or a streamed error.
     */
    suspend fun stream(messages: List<AiMessage>, onDelta: (String) -> Unit) =
        withContext(Dispatchers.IO) {
            val body = JSONObject().apply {
                val arr = JSONArray()
                for (m in messages) {
                    arr.put(JSONObject().apply {
                        put("role", m.role)
                        put("content", m.content)
                    })
                }
                put("messages", arr)
            }

            val conn = (URL("${Config.API_BASE_URL}/api/local/ai/chat").openConnection() as HttpURLConnection).apply {
                requestMethod = "POST"
                connectTimeout = 15_000
                // Token-by-token streams can pause between deltas; keep the read
                // timeout generous so a thinking model doesn't trip it.
                readTimeout = 60_000
                doOutput = true
                setRequestProperty("Content-Type", "application/json")
                setRequestProperty("Accept", "text/event-stream")
            }

            try {
                conn.outputStream.use { out -> out.write(body.toString().toByteArray(Charsets.UTF_8)) }

                val code = conn.responseCode
                if (code !in 200..299) {
                    val text = conn.errorStream?.bufferedReader()?.use { it.readText() }.orEmpty()
                    throw HttpError(code, extractError(text, code))
                }

                conn.inputStream.bufferedReader().use { reader ->
                    reader.lineSequence().forEach { raw ->
                        val line = raw.trim()
                        // SSE frames are blank-line separated; only `data:` lines carry payload.
                        if (!line.startsWith("data:")) return@forEach
                        val payload = line.removePrefix("data:").trim()
                        if (payload.isEmpty()) return@forEach
                        if (payload == "[DONE]") return@use   // graceful end of stream

                        // Each data line is a tiny JSON object: {"delta":…} or {"error":…}.
                        val obj = runCatching { JSONObject(payload) }.getOrNull() ?: return@forEach
                        val streamedError = obj.optString("error")
                        if (streamedError.isNotBlank()) throw HttpError(200, streamedError)
                        val delta = obj.optString("delta")
                        if (delta.isNotEmpty()) onDelta(delta)
                    }
                }
            } finally {
                conn.disconnect()
            }
        }

    private fun extractError(text: String, code: Int): String {
        val parsed = runCatching { JSONObject(text).optString("error") }.getOrNull()
        return if (!parsed.isNullOrBlank()) parsed else "Request failed ($code)"
    }
}

File: mobile/ios/Sources/AITravelChatService.swift
import Foundation

/// One turn in the AI concierge conversation. `role` is `"user"` or
/// `"assistant"`; `content` is the message text. Encoded as the request body's
/// `messages` array and decoded never (the stream returns deltas, not turns).
struct AIMessage: Codable, Equatable {
    let role: String
    var content: String
}

/// Streams the travel-concierge reply from the backend.
///
///   POST {base}/api/local/ai/chat
///   body: { "messages": [ { role, content }, … ] }   (the whole conversation)
///
/// The response is **Server-Sent Events** (`text/event-stream`). We read it line
/// by line with `URLSession.bytes(for:)` and, for each `data:` line:
///   • `{"delta":"…"}`  → hand the text fragment to `onDelta` (append it live)
///   • `{"error":"…"}`  → throw `AIChatError.message(…)`
///   • `[DONE]`         → end of stream, stop
///
/// Auth is optional (public endpoint); we attach the bearer token when present,
/// mirroring `NotificationService`, but a signed-out guest can chat too.
struct AITravelChatService {
    static let shared = AITravelChatService()

    private let session: URLSession = {
        let cfg = URLSessionConfiguration.default
        // Generous request timeout: the model can take a moment before the first
        // token, and the connection then stays open for the whole stream.
        cfg.timeoutIntervalForRequest = 60
        cfg.waitsForConnectivity = true
        return URLSession(configuration: cfg)
    }()

    /// The persisted bearer token, or `nil` when browsing as a guest. Optional
    /// for this endpoint — sending it is fine, omitting it is fine.
    private var token: String? {
        let value = UserDefaults.standard.string(forKey: AuthStore.tokenKey)
        return (value?.isEmpty == false) ? value : nil
    }

    /// Open the stream and feed every text fragment to `onDelta` as it arrives.
    /// Returns once the server sends `[DONE]`. Throws `AIChatError` on a non-2xx
    /// response or an inline `{"error":…}` event. `onDelta` is always invoked on
    /// the main actor so callers can mutate UI state directly.
    func stream(messages: [AIMessage], onDelta: @escaping @MainActor (String) -> Void) async throws {
        guard let url = URL(string: "\(Config.apiBaseURL)/api/local/ai/chat") else {
            throw AIChatError.generic
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("text/event-stream", forHTTPHeaderField: "Accept")
        if let token {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        request.httpBody = try JSONEncoder().encode(RequestBody(messages: messages))

        let (bytes, response) = try await session.bytes(for: request)

        guard let http = response as? HTTPURLResponse else {
            throw AIChatError.generic
        }

        // Non-200: the body is JSON `{"error":"…"}` (e.g. 503 when the key isn't
        // configured). Drain the stream into Data and surface a friendly message.
        guard (200...299).contains(http.statusCode) else {
            var data = Data()
            for try await byte in bytes { data.append(byte) }
            if http.statusCode == 503 {
                throw AIChatError.unavailable
            }
            if let serverMessage = Self.decodeError(data) {
                throw AIChatError.message(serverMessage)
            }
            throw AIChatError.generic
        }

        for try await line in bytes.lines {
            // SSE frames are `data: <payload>`; blank lines separate events.
            let trimmed = line.trimmingCharacters(in: .whitespaces)
            guard trimmed.hasPrefix("data:") else { continue }
            let payload = String(trimmed.dropFirst("data:".count))
                .trimmingCharacters(in: .whitespaces)
            if payload.isEmpty { continue }
            if payload == "[DONE]" { return }

            guard let json = payload.data(using: .utf8) else { continue }
            let event = try? JSONDecoder().decode(StreamEvent.self, from: json)
            if let error = event?.error, !error.isEmpty {
                throw AIChatError.message(error)
            }
            if let delta = event?.delta, !delta.isEmpty {
                await onDelta(delta)
            }
        }
    }

    // MARK: - Wire types

    private struct RequestBody: Encodable {
        let messages: [AIMessage]
    }

    /// A single SSE data frame. Both fields are optional so one decoder handles
    /// `{"delta":…}` and `{"error":…}` frames alike.
    private struct StreamEvent: Decodable {
        let delta: String?
        let error: String?
    }

    private static func decodeError(_ data: Data) -> String? {
        struct ErrorBody: Decodable { let error: String }
        return (try? JSONDecoder().decode(ErrorBody.self, from: data))?.error
    }
}

/// Errors surfaced to the concierge UI.
///
/// The enum is deliberately free of `L.t` lookups (which are `@MainActor`-bound)
/// so the service stays nonisolated, mirroring `NotificationService` /
/// `BookingService`. The view model — already on the main actor — turns each
/// case into a localized string via `localizedMessage`.
enum AIChatError: LocalizedError {
    /// A generic transport/parse failure.
    case generic
    /// The AI service isn't configured yet (HTTP 503) — show a soft note.
    case unavailable
    /// A specific, already-human-readable message from the server.
    case message(String)

    /// Localized text for the inline error bubble. Resolved on the main actor.
    @MainActor var localizedMessage: String {
        switch self {
        case .generic: return L.t("ai.error.generic")
        case .unavailable: return L.t("ai.error.unavailable")
        case let .message(text): return text
        }
    }
}
